dask_cuda/device_host_file.py

```diff
@@ -0,0 +1,96 @@
+from zict import Buffer, File, Func
+from zict.common import ZictBase
+from distributed.protocol import deserialize_bytes, serialize_bytes
+from distributed.worker import weight
+
+from functools import partial
+import os
+
+
+def _is_device_object(obj):
+    """
+    Check if obj is a device object, by checking if it has a
+    __cuda_array_interface__ attributed
+    """
+    return hasattr(obj, '__cuda_array_interface__')
+
+
+def _serialize_if_device(obj):
+    """ Serialize an object if it's a device object """
+    if _is_device_object(obj):
+        return serialize_bytes(obj, on_error='raise')
+    else:
+        return obj
+
+
+def _deserialize_if_device(obj):
+    """ Deserialize an object if it's an instance of bytes """
+    if isinstance(obj, bytes):
+        return deserialize_bytes(obj)
+    else:
+        return obj
+
+
+class DeviceHostFile(ZictBase):
+    """ Manages serialization/deserialization of objects.
+
+    Three LRU cache levels are controlled, for device, host and disk.
+    Each level takes care of serializing objects once its limit has been
+    reached and pass it to the subsequent level. Similarly, each cache
+    may deserialize the object, but storing it back in the appropriate
+    cache, depending on the type of object being deserialized.
+
+    Parameters
+    ----------
+    device_memory_limit: int
+        Number of bytes of CUDA device memory for device LRU cache,
+        spills to host cache once filled.
+    memory_limit: int
+        Number of bytes of host memory for host LRU cache, spills to
+        disk once filled.
+    local_dir: path
+        Path where to store serialized objects on disk
+    """
+    def __init__(self, device_memory_limit=None, memory_limit=None,
+                 local_dir='dask-worker-space'):
+        path = os.path.join(local_dir, 'storage')
+
+        self.host_func = dict()
+        self.disk_func = Func(partial(serialize_bytes, on_error='raise'),
+                              deserialize_bytes, File(path))
+        self.host = Buffer(self.host_func, self.disk_func, memory_limit,
+                           weight=weight)
+
+        self.device_func = dict()
+        self.device_host_func = Func(_serialize_if_device,
+                                     _deserialize_if_device, self.host)
+        self.device = Buffer(self.device_func, self.device_host_func,
+                             device_memory_limit, weight=weight)
+
+        self.fast = self.host.fast
```

Comment:
Thinking a bit about the names here.  My first reaction was to expect `self.device` to include only those pieces of data that were on the device and `self.host` to include only those pieces of data that were on the host.  Currently `self.device` includes everything.

As a result, we might consider a renaming, something like 

```python
self.device = dict()
self.host = dict()
self.disk = Func(..., File(...))
```

And then we might rename the buffers to something else (I can't think of a good name right now).  

This makes it somewhat pleasant when inspecting this object, to see what is on the device and host, and so on.  One can do this now, but only if you know how to dive within `Buffer` objects.

Reply:
I tried doing this in aa0ddd4 but wasn't able to make things clean before I had to context switch away.

Reply:
I'm not sure those changes make things much clearer. FWIW, I'd normally prefer to document what such attributes mean, from a user's perspective, I'm not sure how much clearer things get from renaming `device` to `device_host_disk`. And already thinking a bit into the future, it could get very complex maintaining that information clear, when we do multiple storage media, we'll stack multiple of these Buffers, that may not be very clear to the user how they are organized internally anyway.

Also, is there a use case for users to access that data directly?

Reply:
Yes, people often want to ask things like "how much of my data is currently on the device vs on the host?"  They'll run Dask functions that let them query the worker state directly (this is common among advanced users).  With the current naming they might do something like `len(self.device)` thinking that this is the number of items on the device, while in fact they need to do `len(self.device.fast)` which requires knowledge of how Zict works.

Reply:
Perhaps we should then provide wrappers to that information only? e.g., `self.device` then provides only the information from the device, but that coming from `Buffer` and not from its function (currently, the `dict()`).

Reply:
In other words, `self.device` would provide only `self.device_host_disk.fast` (using the naming scheme from your changes).

Reply:
Yes, that's my primary objective in the renaming.  I mostly want the names to be easily interprettable for novice users who run into this object quickly without knowing much about the object.

Reply:
You'd like to finish that yourself? Otherwise I can do that tomorrow.

Reply:
Nope.  Other things have taken precedence.  I thought it might be a quick fix, so thought it'd be faster to just do it.  I ended up being mistaken :)

Also happy with other naming schemes if you prefer.  Mostly I want to avoid having to explain how to dive into these objects to users in the future.  If documentation is the best way to do that, I'm happy with that as well.

Reply:
I will do it tomorrow then.

I'm fine with the naming scheme, as long as we can keep the external names simple (`device`,  `host` and `disk`), the remaining names are not that much important, as long as they more or less reflect their purpose. I simply wanted to avoid changing names for the purpose of "documenting" code, but since we need to expose internals to the user, that's a good reason. :)

Reply:
I pushed the changes we discussed yesterday, let me know if you agree with the naming scheme.